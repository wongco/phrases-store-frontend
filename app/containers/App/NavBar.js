import React from 'react';
import { NavLink } from 'react-router-dom';
import styled from 'styled-components';

const Nav = styled.nav`
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  background-color: lightgray;
  width: 100%;
  border-bottom-left-radius: 5px;
  border-bottom-right-radius: 5px;
  box-shadow: 2px 2px 6px skyblue;

  /* Permalink - use to edit and share this gradient: http://colorzilla.com/gradient-editor/#1e5799+0,207cca+0,207cca+24,207cca+39,207cca+39,2989d8+73,7db9e8+99 */
  background: #1e5799; /* Old browsers */
  background: -moz-linear-gradient(
    top,
    #1e5799 0%,
    #207cca 0%,
    #207cca 24%,
    #207cca 39%,
    #207cca 39%,
    #2989d8 73%,
    #7db9e8 99%
  ); /* FF3.6-15 */
  background: -webkit-linear-gradient(
    top,
    #1e5799 0%,
    #207cca 0%,
    #207cca 24%,
    #207cca 39%,
    #207cca 39%,
    #2989d8 73%,
    #7db9e8 99%
  ); /* Chrome10-25,Safari5.1-6 */
  background: linear-gradient(
    to bottom,
    #1e5799 0%,
    #207cca 0%,
    #207cca 24%,
    #207cca 39%,
    #207cca 39%,
    #2989d8 73%,
    #7db9e8 99%
  ); /* W3C, IE10+, FF16+, Chrome26+, Opera12+, Safari7+ */
  filter: progid:DXImageTransform.Microsoft.gradient( startColorstr='#1e5799', endColorstr='#7db9e8',GradientType=0 ); /* IE6-9 */
`;

const StyledNavLink = styled(NavLink)`
  text-align: center;
  color: white;
  text-decoration: none;
  padding: 1em;
  margin: 0 2em;
  width: 120px;
  font-weight: 200;
  -webkit-transition: background-color 0.5s; /* Safari */
  transition: background-color 0.5s;

  &:hover {
    background-color: skyblue;
  }
`;

const Hero = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  padding: 0 2em;
  font-weight: bold;
`;

const NavBar = () => {
  const activeStyle = {
    fontWeight: '450',
    color: 'rgb(7, 196, 98)',
  };

  return (
    <div>
      <Nav>
        <StyledNavLink exact to="/" activeStyle={activeStyle}>
          Home
        </StyledNavLink>
        <StyledNavLink exact to="/add" activeStyle={activeStyle}>
          Add
        </StyledNavLink>
        <StyledNavLink exact to="/phrases" activeStyle={activeStyle}>
          Show All
        </StyledNavLink>
        <Hero>The Phrases Store</Hero>
      </Nav>
    </div>
  );
};

export default NavBar;
