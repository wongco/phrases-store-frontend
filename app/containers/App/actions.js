/*
 * App Actions
 *
 * Actions change things in your application
 * Since this boilerplate uses a uni-directional data flow, specifically redux,
 * we have these actions which are the only way your application interacts with
 * your application state. This guarantees that your state is up to date and nobody
 * messes it up weirdly somewhere.
 *
 * To add a new Action:
 * 1) Import your constant
 * 2) Add a function like this:
 *    export function yourAction(var) {
 *        return { type: YOUR_ACTION_CONSTANT, var: var }
 *    }
 */
import {
  LOAD_PHRASES,
  LOAD_PHRASES_SUCCESS,
  LOAD_PHRASES_ERROR,
} from './constants';

/**
 * @description Load the phrases, this action starts the request saga
 * @return {object} An action object with a type of LOAD_PHRASES
 */
export function loadPhrases() {
  return {
    type: LOAD_PHRASES,
  };
}

/**
 * @description Dispatched when the phrases are loaded by the request saga
 * @param  {array} phrases The phrases data
 * @return {object} An action object with a type of LOAD_PHRASES_SUCCESS passing the phrases
 */
export function phrasesLoaded(phrases) {
  return {
    type: LOAD_PHRASES_SUCCESS,
    phrases,
  };
}

/**
 * Dispatched when loading the phrases fails
 * @param  {object} error The error
 * @return {object} An action object with a type of LOAD_PHRASES_ERROR passing the error
 */
export function phrasesLoadingError(error) {
  return {
    type: LOAD_PHRASES_ERROR,
    error,
  };
}
