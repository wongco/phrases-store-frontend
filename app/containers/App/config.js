/* Application config constants */

export const API_BASE_URL = 'http://localhost:5000';
export const API_TIMEOUT_IN_MS = 10000;
