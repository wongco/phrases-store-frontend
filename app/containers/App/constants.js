/*
 * AppConstants
 * Each action has a corresponding type, which the reducer knows and picks up on.
 * To avoid weird typos between the reducer and the actions, we save them as
 * constants here. We prefix them with 'yourproject/YourComponent' so we avoid
 * reducers accidentally picking up actions they shouldn't.
 *
 * Follow this format:
 * export const YOUR_ACTION_CONSTANT = 'yourproject/YourContainer/YOUR_ACTION_CONSTANT';
 */

export const LOAD_PHRASES = 'phrases-stores/App/LOAD_PHRASES';
export const LOAD_PHRASES_SUCCESS = 'phrases-stores/App/LOAD_PHRASES_SUCCESS';
export const LOAD_PHRASES_ERROR = 'phrases-stores/App/LOAD_PHRASES_ERROR';
export const ADD_PHRASE = 'phrases-stores/App/ADD_PHRASE';
export const ADD_PHRASE_SUCCESS = 'phrases-stores/App/ADD_PHRASE_SUCCESS';
export const ADD_PHRASE_ERROR = 'phrases-stores/App/ADD_PHRASE_ERROR';
