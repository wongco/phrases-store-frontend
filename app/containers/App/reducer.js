/*
 * AppReducer
 *
 * The reducer takes care of our data. Using actions, we can change our
 * application state.
 * To add a new action, add it to the switch statement in the reducer function
 *
 * Example:
 * case YOUR_ACTION_CONSTANT:
 *   return state.set('yourStateVariable', true);
 */

import { fromJS } from 'immutable';

import {
  LOAD_PHRASES,
  LOAD_PHRASES_SUCCESS,
  ADD_PHRASE,
  ADD_PHRASE_SUCCESS,
  SET_APP_ERROR,
  CLEAR_ERROR,
} from './constants';

// The initial state of the App
const initialState = fromJS({
  loading: false,
  error: false,
  phrases: [],
});

function appReducer(state = initialState, action) {
  switch (action.type) {
    case LOAD_PHRASES:
      return state.set('loading', true).set('error', false);
    case LOAD_PHRASES_SUCCESS:
      return state.set('phrases', action.phrases).set('loading', false);
    case ADD_PHRASE:
      return state.set('loading', true).set('error', false);
    case ADD_PHRASE_SUCCESS:
      return state.set('loading', false);
    case SET_APP_ERROR:
      return state.set('error', action.error).set('loading', false);
    case CLEAR_ERROR:
      return state.set('error', false);
    default:
      return state;
  }
}

export default appReducer;
