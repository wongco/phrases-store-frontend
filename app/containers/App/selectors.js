/**
 * The global state selectors
 */

import { createSelector } from 'reselect';

const selectGlobal = state => state.get('global');

const selectRouter = state => state.get('router');

const makeSelectLoading = () =>
  createSelector(selectGlobal, globalState => globalState.get('loading'));

const makeSelectError = () =>
  createSelector(selectGlobal, globalState => globalState.get('error'));

const makeSelectPhrases = () =>
  createSelector(selectGlobal, globalState => globalState.get('phrases'));

const makeSelectLocation = () =>
  createSelector(selectRouter, routerState =>
    routerState.get('location').toJS(),
  );

export {
  selectGlobal,
  makeSelectLoading,
  makeSelectError,
  makeSelectPhrases,
  makeSelectLocation,
};
