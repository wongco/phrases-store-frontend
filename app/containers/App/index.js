/**
 *
 * App.js
 *
 * This component is the skeleton around the actual pages, and should only
 * contain code that should be seen on all pages. (e.g. navigation bar)
 *
 */

import React from 'react';
import { Switch, Route } from 'react-router-dom';
import { Helmet } from 'react-helmet';
import styled from 'styled-components';

import HomePage from 'containers/HomePage/Loadable';
import NotFoundPage from 'containers/NotFoundPage/Loadable';
import PhrasesView from 'containers/PhrasesView/Loadable';
import AddPhrase from 'containers/AddPhrase/Loadable';
import NavBar from './NavBar';

import GlobalStyle from '../../global-styles';

const AppWrapper = styled.div`
  display: flex;
  flex-direction: column;
  min-height: 100%;
  width: calc(768px + 1em * 2);
  padding: 0 1em;
  margin: 0 auto;
`;

export default function App() {
  return (
    <AppWrapper>
      <Helmet>
        <title>The Phrases Store</title>
        <meta name="description" content="a phrase storing application" />
      </Helmet>
      <NavBar />
      <Switch>
        <Route exact path="/" component={HomePage} />
        <Route exact path="/phrases" component={PhrasesView} />
        <Route exact path="/add" component={AddPhrase} />
        <Route component={NotFoundPage} />
      </Switch>
      <GlobalStyle />
    </AppWrapper>
  );
}
