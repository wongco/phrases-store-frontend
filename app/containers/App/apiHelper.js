import axios from 'axios';
import { API_BASE_URL, API_TIMEOUT_IN_MS } from './config';

function baseApiRequestData() {
  return {
    url: `${API_BASE_URL}/phrases`,
    timeout: API_TIMEOUT_IN_MS,
  };
}

/**
 * @description shape apiCall request data for getting phrases
 * @return { object} GET request object
 */
export function createGetRequestObj() {
  return {
    method: 'get',
    ...baseApiRequestData(),
  };
}

/**
 * @description shape apiCall request data for adding phrase
 * @param { string } text - user input phrase to add to API
 * @return { object } POST request object
 */
export function createPostRequestObj(text) {
  return {
    method: 'post',
    data: {
      phrase: {
        text,
      },
    },
    ...baseApiRequestData(),
  };
}

/**
 * @description makes api call with provided requestObj
 * @param { object } requestData - API request data incl. method and body/query
 * @return { Promsie <apiResponse> } - Promise of API Response
 */
export function apiCall(requestObj) {
  return axios(requestObj);
}
