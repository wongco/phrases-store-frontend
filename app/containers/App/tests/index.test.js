import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Route } from 'react-router-dom';

import GlobalStyle from '../../../global-styles';
import NavBar from '../NavBar';
import App from '../index';

describe('<App />', () => {
  it('renders without crashing', () => {
    shallow(<App />);
  });

  it('matches snapshot for default criteria', () => {
    const wrapper = shallow(<App />);
    const serialized = toJson(wrapper);
    expect(serialized).toMatchSnapshot();
  });

  it('should render the navbar', () => {
    const renderedComponent = shallow(<App />);
    expect(renderedComponent.find(NavBar)).toHaveLength(1);
  });

  it('should render some routes', () => {
    const renderedComponent = shallow(<App />);
    expect(renderedComponent.find(Route).length).not.toBe(0);
  });

  it('should render global styles', () => {
    const renderedComponent = shallow(<App />);
    expect(renderedComponent.find(GlobalStyle)).toHaveLength(1);
  });
});
