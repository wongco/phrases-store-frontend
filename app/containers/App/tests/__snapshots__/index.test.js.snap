// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> matches snapshot for default criteria 1`] = `
<App__AppWrapper>
  <HelmetWrapper
    defer={true}
    encodeSpecialCharacters={true}
  >
    <title>
      The Phrases Store
    </title>
    <meta
      content="a phrase storing application"
      name="description"
    />
  </HelmetWrapper>
  <NavBar />
  <Switch>
    <Route
      component={[Function]}
      exact={true}
      path="/"
    />
    <Route
      component={[Function]}
      exact={true}
      path="/phrases"
    />
    <Route
      component={[Function]}
      exact={true}
      path="/add"
    />
    <Route
      component={[Function]}
    />
  </Switch>
  <GlobalStyleComponent
    suppressMultiMountWarning={false}
  />
</App__AppWrapper>
`;
